File: src/recurrence.ts
import { canonicalizeMerchant } from "./importer";
import type { ExpenseEntry, IncomeEntry, LedgerState, MonthBudget } from "./types";

// Local recurring-transaction detection — no server, works in privacy mode.
// Groups ledger entries by canonical merchant and looks for a steady cadence
// with a steady amount. Powers the subscriptions panel, "mark as recurring"
// suggestions, missed-payment signals, and the seeded↔imported reconciliation.

export type RecurrenceCadence = "weekly" | "4-weekly" | "monthly" | "quarterly" | "annual";

export type RecurrenceEntryRef = {
  monthKey: string;
  entryId: string;
  kind: "income" | "expense";
};

export type RecurrenceCandidate = {
  key: string;
  label: string;
  kind: "income" | "expense";
  category: string;
  color: string;
  cadence: RecurrenceCadence;
  typicalAmount: number;
  monthlyEquivalent: number;
  occurrences: number;
  lastSeen: string; // ISO date or monthKey
  nextExpected: string | null; // ISO date; null when occurrences carry no dates
  confidence: number;
  alreadyRecurring: boolean;
  entryRefs: RecurrenceEntryRef[];
};

type Occurrence = {
  monthKey: string;
  entryId: string;
  kind: "income" | "expense";
  amount: number;
  date?: string;
  recurring: boolean;
  category: string;
  color: string;
  label: string;
};

const CADENCE_DAYS: Record<RecurrenceCadence, number> = {
  weekly: 7,
  "4-weekly": 28,
  monthly: 30,
  quarterly: 91,
  annual: 365,
};

const MONTHLY_FACTOR: Record<RecurrenceCadence, number> = {
  weekly: 52 / 12,
  "4-weekly": 13 / 12,
  monthly: 1,
  quarterly: 1 / 3,
  annual: 1 / 12,
};

export function detectRecurringCandidates(state: LedgerState): RecurrenceCandidate[] {
  const groups = new Map<string, Occurrence[]>();

  for (const [monthKey, month] of Object.entries(state.months)) {
    for (const income of month.incomes) {
      pushOccurrence(groups, monthKey, income.id, "income", income.amount, income.date, income.recurring, "", income.color, income.source);
    }
    for (const expense of month.expenses) {
      pushOccurrence(groups, monthKey, expense.id, "expense", expense.amount, expense.date, expense.recurring, expense.category, expense.color, expense.name);
    }
  }

  const candidates: RecurrenceCandidate[] = [];
  for (const occurrences of groups.values()) {
    const candidate = classifyGroup(occurrences);
    if (candidate) candidates.push(candidate);
  }

  return candidates.sort((a, b) => b.monthlyEquivalent - a.monthlyEquivalent);
}

// Detected-but-unflagged candidates: the "mark as recurring?" suggestion list.
export function suggestRecurringFlags(state: LedgerState): RecurrenceCandidate[] {
  return detectRecurringCandidates(state).filter((candidate) => !candidate.alreadyRecurring && candidate.confidence >= 0.6);
}

// Expense candidates on a repeating cadence — the subscriptions/bills panel.
export function detectSubscriptions(state: LedgerState): RecurrenceCandidate[] {
  return detectRecurringCandidates(state).filter(
    (candidate) => candidate.kind === "expense" && candidate.confidence >= 0.6,
  );
}

// Candidates whose next expected date has passed with no matching entry — the
// "council tax usually lands by the 3rd" signal. Grace of 4 days before flagging.
export function findMissedRecurring(state: LedgerState, todayIso: string): RecurrenceCandidate[] {
  const currentMonthKey = todayIso.slice(0, 7);
  return detectRecurringCandidates(state).filter((candidate) => {
    if (!candidate.nextExpected || candidate.confidence < 0.7) return false;
    if (candidate.cadence === "weekly") return false; // too noisy to nag about
    const graceCutoff = shiftIsoDate(candidate.nextExpected, 4);
    if (todayIso <= graceCutoff) return false;
    // Not missed if something already landed for it in the expected month.
    const expectedMonth = candidate.nextExpected.slice(0, 7);
    if (expectedMonth < currentMonthKey) return false;
    return !candidate.entryRefs.some((ref) => ref.monthKey === expectedMonth);
  });
}

// A CSV/feed import can bring in the real salary or rent line for a month that
// auto-seeding already populated. Rather than doubling, the seeded copy folds
// into the imported actual: the import wins (real amount + metadata) and takes
// over the recurring role. Returns how many rows merged so the caller can toast.
export function reconcileSeededEntries(month: MonthBudget): { month: MonthBudget; merged: number } {
  let merged = 0;

  const importedIncomeKeys = new Set(
    month.incomes.filter((income) => !income.seededFrom && income.imported).map((income) => canonicalizeMerchant(income.source)),
  );
  const importedExpenseKeys = new Set(
    month.expenses.filter((expense) => !expense.seededFrom && expense.imported).map((expense) => canonicalizeMerchant(expense.name)),
  );

  const seededIncomeByKey = new Map(
    month.incomes.filter((income) => income.seededFrom).map((income) => [canonicalizeMerchant(income.source), income]),
  );
  const seededExpenseByKey = new Map(
    month.expenses.filter((expense) => expense.seededFrom).map((expense) => [canonicalizeMerchant(expense.name), expense]),
  );

  const incomes = month.incomes
    .filter((income) => {
      const drop = Boolean(income.seededFrom && importedIncomeKeys.has(canonicalizeMerchant(income.source)));
      if (drop) merged += 1;
      return !drop;
    })
    .map((income) => {
      if (income.seededFrom || !income.imported) return income;
      const seeded = seededIncomeByKey.get(canonicalizeMerchant(income.source));
      return seeded ? { ...income, recurring: true } : income;
    });

  const expenses = month.expenses
    .filter((expense) => {
      const drop = Boolean(expense.seededFrom && importedExpenseKeys.has(canonicalizeMerchant(expense.name)));
      if (drop) merged += 1;
      return !drop;
    })
    .map((expense) => {
      if (expense.seededFrom || !expense.imported) return expense;
      const seeded = seededExpenseByKey.get(canonicalizeMerchant(expense.name));
      if (!seeded) return expense;
      // The seeded copy knew the category (it came from the recurring template);
      // an Unsorted import inherits it instead of demanding a re-sort.
      return {
        ...expense,
        recurring: true,
        category: expense.category === "Unsorted" && seeded.category ? seeded.category : expense.category,
        color: expense.category === "Unsorted" && seeded.category ? seeded.color : expense.color,
      };
    });

  return merged ? { month: { ...month, incomes, expenses }, merged } : { month, merged: 0 };
}

// Back-fill the recurring flag on every entry that belongs to a candidate.
export function applyRecurringFlag(state: LedgerState, candidate: RecurrenceCandidate): LedgerState {
  const refKeys = new Set(candidate.entryRefs.map((ref) => `${ref.monthKey}:${ref.entryId}`));
  const months = Object.fromEntries(
    Object.entries(state.months).map(([monthKey, month]) => [
      monthKey,
      {
        ...month,
        incomes: month.incomes.map((income: IncomeEntry) =>
          refKeys.has(`${monthKey}:${income.id}`) ? { ...income, recurring: true } : income,
        ),
        expenses: month.expenses.map((expense: ExpenseEntry) =>
          refKeys.has(`${monthKey}:${expense.id}`) ? { ...expense, recurring: true } : expense,
        ),
      },
    ]),
  );
  return { ...state, months };
}

function pushOccurrence(
  groups: Map<string, Occurrence[]>,
  monthKey: string,
  entryId: string,
  kind: "income" | "expense",
  amount: number,
  date: string | undefined,
  recurring: boolean,
  category: string,
  color: string,
  label: string,
) {
  const key = canonicalizeMerchant(label);
  if (!key || amount <= 0) return;
  const groupKey = `${kind}:${key}`;
  const list = groups.get(groupKey) ?? [];
  list.push({ monthKey, entryId, kind, amount, date, recurring, category, color, label });
  groups.set(groupKey, list);
}

function classifyGroup(occurrences: Occurrence[]): RecurrenceCandidate | null {
  if (occurrences.length < 3) return null;

  const sorted = [...occurrences].sort((a, b) => (a.date ?? a.monthKey).localeCompare(b.date ?? b.monthKey));
  const amounts = sorted.map((occurrence) => occurrence.amount);
  const typicalAmount = median(amounts.slice(-3));
  const tolerance = Math.max(2, typicalAmount * 0.15);
  const amountsStable = amounts.filter((amount) => Math.abs(amount - typicalAmount) <= tolerance).length >= sorted.length - 1;
  if (!amountsStable) return null;

  const dated = sorted.filter((occurrence) => occurrence.date);
  let cadence: RecurrenceCadence | null = null;
  let gapStability = 0;

  if (dated.length >= 3) {
    const gaps: number[] = [];
    for (let index = 1; index < dated.length; index += 1) {
      gaps.push(dayDiff(dated[index - 1].date!, dated[index].date!));
    }
    const medianGap = median(gaps);
    cadence = cadenceFromGap(medianGap);
    if (cadence) {
      const window = cadence === "weekly" ? 2 : 4;
      const inWindow = gaps.filter((gap) => Math.abs(gap - medianGap) <= window).length;
      gapStability = inWindow / gaps.length;
      if (gapStability < 0.6) cadence = null;
    }
  }

  if (!cadence) {
    // Undated (manual) entries: appearing in 3+ distinct, mostly-consecutive months
    // with a stable amount is a monthly pattern.
    const months = Array.from(new Set(sorted.map((occurrence) => occurrence.monthKey))).sort();
    if (months.length < 3) return null;
    const consecutive = countConsecutiveMonthPairs(months);
    if (consecutive < months.length - 2) return null;
    cadence = "monthly";
    gapStability = consecutive / (months.length - 1);
  }

  const last = sorted[sorted.length - 1];
  const lastSeen = last.date ?? last.monthKey;
  const nextExpected = last.date ? shiftIsoDate(last.date, CADENCE_DAYS[cadence]) : null;

  const confidence = Math.min(
    0.95,
    0.55 + 0.05 * Math.min(sorted.length, 7) + 0.2 * gapStability,
  );

  const latestCategorized = [...sorted].reverse().find((occurrence) => occurrence.category);
  return {
    key: canonicalizeMerchant(last.label),
    label: last.label,
    kind: last.kind,
    category: latestCategorized?.category ?? "",
    color: last.color,
    cadence,
    typicalAmount,
    monthlyEquivalent: Number((typicalAmount * MONTHLY_FACTOR[cadence]).toFixed(2)),
    occurrences: sorted.length,
    lastSeen,
    nextExpected,
    confidence: Number(confidence.toFixed(2)),
    alreadyRecurring: sorted.every((occurrence) => occurrence.recurring),
    entryRefs: sorted.map((occurrence) => ({ monthKey: occurrence.monthKey, entryId: occurrence.entryId, kind: occurrence.kind })),
  };
}

function cadenceFromGap(medianGap: number): RecurrenceCadence | null {
  if (medianGap >= 5 && medianGap <= 9) return "weekly";
  // 4-weekly only when the gap sits tightly at 27–29; day-of-month billing drifts 28–33.
  if (medianGap >= 26 && medianGap < 29.5) return "4-weekly";
  if (medianGap >= 29.5 && medianGap <= 33) return "monthly";
  if (medianGap >= 85 && medianGap <= 97) return "quarterly";
  if (medianGap >= 350 && medianGap <= 380) return "annual";
  return null;
}

function countConsecutiveMonthPairs(sortedMonths: string[]): number {
  let pairs = 0;
  for (let index = 1; index < sortedMonths.length; index += 1) {
    if (monthDiff(sortedMonths[index - 1], sortedMonths[index]) === 1) pairs += 1;
  }
  return pairs;
}

function monthDiff(a: string, b: string): number {
  const [ay, am] = a.split("-").map(Number);
  const [by, bm] = b.split("-").map(Number);
  return (by - ay) * 12 + (bm - am);
}

function dayDiff(a: string, b: string): number {
  return Math.round((Date.parse(b) - Date.parse(a)) / 86_400_000);
}

function shiftIsoDate(iso: string, days: number): string {
  const date = new Date(`${iso}T00:00:00`);
  date.setDate(date.getDate() + days);
  return `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, "0")}-${String(date.getDate()).padStart(2, "0")}`;
}

function median(values: number[]): number {
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[middle] : (sorted[middle - 1] + sorted[middle]) / 2;
}

File: ios/IncomeTracker/Engine/Recurrence.swift
// Recurrence.swift
// Ported from src/recurrence.ts — Swift 6, Foundation only.
//
// Local recurring-transaction detection — no server, works in privacy mode.
// Groups ledger entries by canonical merchant and looks for a steady cadence
// with a steady amount. Powers the subscriptions panel, "mark as recurring"
// suggestions, missed-payment signals, and the seeded↔imported reconciliation.

import Foundation

public enum RecurrenceCadence: String, Sendable, Codable, Hashable {
    case weekly
    case fourWeekly = "4-weekly"
    case monthly
    case quarterly
    case annual

    var days: Int {
        switch self {
        case .weekly: return 7
        case .fourWeekly: return 28
        case .monthly: return 30
        case .quarterly: return 91
        case .annual: return 365
        }
    }

    var monthlyFactor: Double {
        switch self {
        case .weekly: return 52.0 / 12.0
        case .fourWeekly: return 13.0 / 12.0
        case .monthly: return 1
        case .quarterly: return 1.0 / 3.0
        case .annual: return 1.0 / 12.0
        }
    }

    public var label: String {
        switch self {
        case .weekly: return "Weekly"
        case .fourWeekly: return "Every 4 weeks"
        case .monthly: return "Monthly"
        case .quarterly: return "Quarterly"
        case .annual: return "Yearly"
        }
    }
}

public struct RecurrenceEntryRef: Sendable, Hashable {
    public let monthKey: String
    public let entryId: String
    public let kind: TransactionKind
}

public struct RecurrenceCandidate: Sendable, Hashable, Identifiable {
    public var id: String { "\(kind.rawValue):\(key)" }
    public let key: String
    public let label: String
    public let kind: TransactionKind
    public let category: String
    public let color: String
    public let cadence: RecurrenceCadence
    public let typicalAmount: Double
    public let monthlyEquivalent: Double
    public let occurrences: Int
    /// ISO date or monthKey.
    public let lastSeen: String
    /// ISO date; nil when occurrences carry no dates.
    public let nextExpected: String?
    public let confidence: Double
    public let alreadyRecurring: Bool
    public let entryRefs: [RecurrenceEntryRef]
}

private struct Occurrence {
    let monthKey: String
    let entryId: String
    let kind: TransactionKind
    let amount: Double
    let date: String?
    let recurring: Bool
    let category: String
    let color: String
    let label: String

    var sortKey: String { date ?? monthKey }
}

public enum Recurrence {

    public static func detectCandidates(_ state: LedgerState) -> [RecurrenceCandidate] {
        var groups: [String: [Occurrence]] = [:]
        var order: [String] = []

        for monthKey in state.months.keys.sorted() {
            guard let month = state.months[monthKey] else { continue }
            for income in month.incomes {
                push(&groups, &order, Occurrence(monthKey: monthKey, entryId: income.id, kind: .income, amount: income.amount,
                                                  date: income.date, recurring: income.recurring, category: "",
                                                  color: income.color, label: income.source))
            }
            for expense in month.expenses {
                push(&groups, &order, Occurrence(monthKey: monthKey, entryId: expense.id, kind: .expense, amount: expense.amount,
                                                  date: expense.date, recurring: expense.recurring, category: expense.category,
                                                  color: expense.color, label: expense.name))
            }
        }

        return order.compactMap { groups[$0] }.compactMap(classify)
            .sorted { $0.monthlyEquivalent > $1.monthlyEquivalent }
    }

    /// Detected-but-unflagged candidates: the "mark as recurring?" suggestion list.
    public static func suggestRecurringFlags(_ state: LedgerState) -> [RecurrenceCandidate] {
        detectCandidates(state).filter { !$0.alreadyRecurring && $0.confidence >= 0.6 }
    }

    /// Expense candidates on a repeating cadence — the subscriptions/bills panel.
    public static func detectSubscriptions(_ state: LedgerState) -> [RecurrenceCandidate] {
        detectCandidates(state).filter { $0.kind == .expense && $0.confidence >= 0.6 }
    }

    /// Candidates whose next expected date has passed with no matching entry — the
    /// "council tax usually lands by the 3rd" signal. Grace of 4 days before flagging.
    public static func findMissed(_ state: LedgerState, todayIso: String) -> [RecurrenceCandidate] {
        let currentMonthKey = String(todayIso.prefix(7))
        return detectCandidates(state).filter { candidate in
            guard let nextExpected = candidate.nextExpected, candidate.confidence >= 0.7 else { return false }
            if candidate.cadence == .weekly { return false }
            let graceCutoff = shiftIsoDate(nextExpected, days: 4)
            if todayIso <= graceCutoff { return false }
            let expectedMonth = String(nextExpected.prefix(7))
            if expectedMonth < currentMonthKey { return false }
            return !candidate.entryRefs.contains { $0.monthKey == expectedMonth }
        }
    }

    /// A CSV/feed import can bring in the real salary or rent line for a month that
    /// auto-seeding already populated. Rather than doubling, the seeded copy folds
    /// into the imported actual: the import wins (real amount + metadata) and takes
    /// over the recurring role. Returns how many rows merged so the caller can toast.
    public static func reconcileSeededEntries(_ month: MonthBudget) -> (month: MonthBudget, merged: Int) {
        var merged = 0

        let importedIncomeKeys = Set(month.incomes.filter { $0.seededFrom == nil && $0.imported != nil }.map { canonicalizeMerchant($0.source) })
        let importedExpenseKeys = Set(month.expenses.filter { $0.seededFrom == nil && $0.imported != nil }.map { canonicalizeMerchant($0.name) })

        var seededIncomeByKey: [String: IncomeEntry] = [:]
        for income in month.incomes where income.seededFrom != nil {
            seededIncomeByKey[canonicalizeMerchant(income.source)] = income
        }
        var seededExpenseByKey: [String: ExpenseEntry] = [:]
        for expense in month.expenses where expense.seededFrom != nil {
            seededExpenseByKey[canonicalizeMerchant(expense.name)] = expense
        }

        let incomes: [IncomeEntry] = month.incomes.compactMap { income in
            if income.seededFrom != nil, importedIncomeKeys.contains(canonicalizeMerchant(income.source)) {
                merged += 1
                return nil
            }
            guard income.seededFrom == nil, income.imported != nil,
                  seededIncomeByKey[canonicalizeMerchant(income.source)] != nil else { return income }
            var updated = income
            updated.recurring = true
            return updated
        }

        let expenses: [ExpenseEntry] = month.expenses.compactMap { expense in
            if expense.seededFrom != nil, importedExpenseKeys.contains(canonicalizeMerchant(expense.name)) {
                merged += 1
                return nil
            }
            guard expense.seededFrom == nil, expense.imported != nil,
                  let seeded = seededExpenseByKey[canonicalizeMerchant(expense.name)] else { return expense }
            var updated = expense
            updated.recurring = true
            // The seeded copy knew the category (it came from the recurring template);
            // an Unsorted import inherits it instead of demanding a re-sort.
            if expense.category == "Unsorted", !seeded.category.isEmpty {
                updated.category = seeded.category
                updated.color = seeded.color
            }
            return updated
        }

        guard merged > 0 else { return (month, 0) }
        var result = month
        result.incomes = incomes
        result.expenses = expenses
        return (result, merged)
    }

    /// Back-fill the recurring flag on every entry that belongs to a candidate.
    public static func applyRecurringFlag(_ state: LedgerState, candidate: RecurrenceCandidate) -> LedgerState {
        let refKeys = Set(candidate.entryRefs.map { "\($0.monthKey):\($0.entryId)" })
        var next = state
        for (monthKey, month) in state.months {
            var updated = month
            updated.incomes = month.incomes.map { income in
                guard refKeys.contains("\(monthKey):\(income.id)") else { return income }
                var copy = income
                copy.recurring = true
                return copy
            }
            updated.expenses = month.expenses.map { expense in
                guard refKeys.contains("\(monthKey):\(expense.id)") else { return expense }
                var copy = expense
                copy.recurring = true
                return copy
            }
            next.months[monthKey] = updated
        }
        return next
    }

    // MARK: - Private

    private static func push(_ groups: inout [String: [Occurrence]], _ order: inout [String], _ occurrence: Occurrence) {
        let key = canonicalizeMerchant(occurrence.label)
        guard !key.isEmpty, occurrence.amount > 0 else { return }
        let groupKey = "\(occurrence.kind.rawValue):\(key)"
        if groups[groupKey] == nil { order.append(groupKey) }
        groups[groupKey, default: []].append(occurrence)
    }

    private static func classify(_ occurrences: [Occurrence]) -> RecurrenceCandidate? {
        guard occurrences.count >= 3 else { return nil }

        let sorted = occurrences.sorted { $0.sortKey < $1.sortKey }
        let amounts = sorted.map(\.amount)
        let typicalAmount = median(Array(amounts.suffix(3)))
        let tolerance = max(2, typicalAmount * 0.15)
        let stableCount = amounts.filter { abs($0 - typicalAmount) <= tolerance }.count
        guard stableCount >= sorted.count - 1 else { return nil }

        let dated = sorted.filter { $0.date != nil }
        var cadence: RecurrenceCadence?
        var gapStability = 0.0

        if dated.count >= 3 {
            var gaps: [Double] = []
            for index in 1..<dated.count {
                gaps.append(Double(dayDiff(dated[index - 1].date!, dated[index].date!)))
            }
            let medianGap = median(gaps)
            cadence = cadenceFromGap(medianGap)
            if let resolved = cadence {
                let window = resolved == .weekly ? 2.0 : 4.0
                let inWindow = gaps.filter { abs($0 - medianGap) <= window }.count
                gapStability = Double(inWindow) / Double(gaps.count)
                if gapStability < 0.6 { cadence = nil }
            }
        }

        if cadence == nil {
            // Undated (manual) entries: appearing in 3+ distinct, mostly-consecutive months
            // with a stable amount is a monthly pattern.
            let months = Array(Set(sorted.map(\.monthKey))).sorted()
            guard months.count >= 3 else { return nil }
            let consecutive = countConsecutiveMonthPairs(months)
            guard consecutive >= months.count - 2 else { return nil }
            cadence = .monthly
            gapStability = Double(consecutive) / Double(months.count - 1)
        }

        guard let resolvedCadence = cadence, let last = sorted.last else { return nil }
        let lastSeen = last.date ?? last.monthKey
        let nextExpected = last.date.map { shiftIsoDate($0, days: resolvedCadence.days) }
        let confidence = min(0.95, 0.55 + 0.05 * Double(min(sorted.count, 7)) + 0.2 * gapStability)
        let latestCategorised = sorted.reversed().first { !$0.category.isEmpty }

        return RecurrenceCandidate(
            key: canonicalizeMerchant(last.label),
            label: last.label,
            kind: last.kind,
            category: latestCategorised?.category ?? "",
            color: last.color,
            cadence: resolvedCadence,
            typicalAmount: typicalAmount,
            monthlyEquivalent: roundToPence(typicalAmount * resolvedCadence.monthlyFactor),
            occurrences: sorted.count,
            lastSeen: lastSeen,
            nextExpected: nextExpected,
            confidence: (confidence * 100).rounded() / 100,
            alreadyRecurring: sorted.allSatisfy(\.recurring),
            entryRefs: sorted.map { RecurrenceEntryRef(monthKey: $0.monthKey, entryId: $0.entryId, kind: $0.kind) }
        )
    }

    private static func cadenceFromGap(_ medianGap: Double) -> RecurrenceCadence? {
        if medianGap >= 5 && medianGap <= 9 { return .weekly }
        // 4-weekly only when the gap sits tightly at 27–29; day-of-month billing drifts 28–33.
        if medianGap >= 26 && medianGap < 29.5 { return .fourWeekly }
        if medianGap >= 29.5 && medianGap <= 33 { return .monthly }
        if medianGap >= 85 && medianGap <= 97 { return .quarterly }
        if medianGap >= 350 && medianGap <= 380 { return .annual }
        return nil
    }

    private static func countConsecutiveMonthPairs(_ sortedMonths: [String]) -> Int {
        var pairs = 0
        for index in 1..<sortedMonths.count where monthDiff(sortedMonths[index - 1], sortedMonths[index]) == 1 {
            pairs += 1
        }
        return pairs
    }

    private static func monthDiff(_ a: String, _ b: String) -> Int {
        let pa = a.split(separator: "-").compactMap { Int($0) }
        let pb = b.split(separator: "-").compactMap { Int($0) }
        guard pa.count >= 2, pb.count >= 2 else { return 0 }
        return (pb[0] - pa[0]) * 12 + (pb[1] - pa[1])
    }

    private static func dayDiff(_ a: String, _ b: String) -> Int {
        guard let da = utcDate(a), let db = utcDate(b) else { return 0 }
        return Int((db.timeIntervalSince(da) / 86_400).rounded())
    }

    static func shiftIsoDate(_ iso: String, days: Int) -> String {
        guard let date = utcDate(iso) else { return iso }
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(secondsFromGMT: 0)!
        guard let shifted = cal.date(byAdding: .day, value: days, to: date) else { return iso }
        let comps = cal.dateComponents([.year, .month, .day], from: shifted)
        return String(format: "%04d-%02d-%02d", comps.year ?? 0, comps.month ?? 0, comps.day ?? 0)
    }

    private static func utcDate(_ iso: String) -> Date? {
        let parts = iso.prefix(10).split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3 else { return nil }
        var comps = DateComponents()
        comps.year = parts[0]; comps.month = parts[1]; comps.day = parts[2]
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(secondsFromGMT: 0)!
        return cal.date(from: comps)
    }

    private static func median(_ values: [Double]) -> Double {
        guard !values.isEmpty else { return 0 }
        let sorted = values.sorted()
        let middle = sorted.count / 2
        return sorted.count % 2 == 1 ? sorted[middle] : (sorted[middle - 1] + sorted[middle]) / 2
    }
}
